go/ephemeral/device_ek_storage.go

```diff
@@ -0,0 +1,145 @@
+package ephemeral
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+)
+
+const deviceEKPrefix = "device-eks/device-ephemeral-key"
+
+type DeviceEKStorage struct {
+	sync.Mutex
+	indexOnce *sync.Once
+	storage   ErasableKVStore
+	cache     map[keybase1.EkGeneration]keybase1.DeviceEk
+	keyPrefix string
+}
+
+func NewDeviceEKStorage(g *libkb.GlobalContext) *DeviceEKStorage {
+	keyPrefix := fmt.Sprintf("%s-%s", deviceEKPrefix, g.Env.GetUsername().String())
+	return &DeviceEKStorage{
+		storage:   NewFileErasableKVStore(g),
+		cache:     make(map[keybase1.EkGeneration]keybase1.DeviceEk),
+		indexOnce: new(sync.Once),
+		keyPrefix: keyPrefix,
+	}
+}
+
+func (s *DeviceEKStorage) key(generation keybase1.EkGeneration) string {
+	return fmt.Sprintf("%s-%d.ek", s.keyPrefix, generation)
+}
+
+func (s *DeviceEKStorage) Put(ctx context.Context, generation keybase1.EkGeneration, deviceEK keybase1.DeviceEk) (err error) {
+	s.Lock()
+	defer s.Unlock()
+
+	key := s.key(generation)
+	err = s.storage.Put(ctx, key, deviceEK)
+	if err != nil {
+		return err
+	}
+
+	// cache the result
+	s.cache[generation] = deviceEK
+	return nil
+}
+
+func (s *DeviceEKStorage) Get(ctx context.Context, generation keybase1.EkGeneration) (deviceEK keybase1.DeviceEk, err error) {
+	s.Lock()
+	defer s.Unlock()
+	return s.get(ctx, generation)
+}
+
+func (s *DeviceEKStorage) get(ctx context.Context, generation keybase1.EkGeneration) (deviceEK keybase1.DeviceEk, err error) {
+	deviceEK, ok := s.cache[generation]
+	if ok {
+		return deviceEK, nil
+	}
+
+	key := s.key(generation)
+	data, err := s.storage.Get(ctx, key)
+	if err != nil {
+		return deviceEK, err
+	}
+
+	deviceEK, ok = data.(keybase1.DeviceEk)
+	if !ok {
+		return deviceEK, fmt.Errorf("Unable to cast data to deviceEK")
+	}
+
+	// cache the result
+	s.cache[generation] = deviceEK
+	return deviceEK, nil
+}
+
+func (s *DeviceEKStorage) Delete(generation keybase1.EkGeneration) error {
+	s.Lock()
+	defer s.Unlock()
+
+	// clear the cache
+	delete(s.cache, generation)
+	key := s.key(generation)
+	return s.storage.Erase(key)
+}
+
+func (s *DeviceEKStorage) index(ctx context.Context) (err error) {
+	s.indexOnce.Do(func() {
+		keys, err := s.storage.AllKeys()
+		if err != nil {
+			return
+		}
+		for _, key := range keys {
+			key = strings.TrimSuffix(key, filepath.Ext(key))
+			if strings.HasPrefix(key, deviceEKPrefix) {
+				parts := strings.Split(key, deviceEKPrefix)
+				g, err := strconv.ParseUint(parts[1], 10, 64)
```

Comment:
won't this crash if only one part is returned?

Reply:
we'll always get two parts since i check `HasPrefix` before splitting.

Reply:
Ok, then just add a comment...